### DISCUSSION TITLE
How to introduce naive-ui ?

### DISCUSSION THREAD

**Discussion by zedo75** [vuepress/core#682] [Q&A] [answered]
I tried to introduce [naive-ui](https://www.naiveui.com/) in my vuepress with its default theme. It worked only when hot-updated, namely 
 after installing the dependency in a running progress, but it went into error after restarting the server:

```
Uncaught (in promise) TypeError: Cannot destructure property 'styles' of 'ssrContext' as it is undefined.
    at Object.ssrAdapter [as adapter] (index.js:12)
    at mount (mount.js:54)
    at Object.wrappedMount [as mount] (c.js:11)
    at mountStyle (use-theme.js:16)
    at useTheme (use-theme.js:33)
    at setup (Button.js:133)
    at callWithErrorHandling (runtime-core.esm-bundler.js:155)
    at setupStatefulComponent (runtime-core.esm-bundler.js:7045)
    at setupComponent (runtime-core.esm-bundler.js:7001)
    at mountComponent (runtime-core.esm-bundler.js:4922)
```
My relevant codes are :

markdown
``` md
<ClientOnly>
<CustomComponent/>
</ClientOnly>
```
vue component (registered in `clientAppEnhance.ts`):
``` vue
<template>
    <ClientOnly>
        <n-button>button</n-button>
    </ClientOnly>
</template>
<script lang="ts" setup>
import {NButton} from "naive-ui";
</script>
```

my `vuepress info` is :

```
  System:
    OS: Windows 10 10.0.19043
    CPU: (8) x64 Intel(R) Core(TM) i5-9300H CPU @ 2.40GHz
    Memory: 2.69 GB / 11.85 GB
  Binaries:
    Node: 16.13.0 - E:\Envs\nodejs\node.EXE
    Yarn: 1.22.17 - E:\Envs\node\node_global\yarn.CMD
    npm: 8.1.0 - E:\Envs\nodejs\npm.CMD
  Utilities:
    Git: 2.34.1.
  Browsers:
    Chrome: Not Found
    Edge: Spartan (44.19041.1266.0)
  npmPackages:
    @vuepress/bundler-vite:  2.0.0-beta.35
    @vuepress/bundler-webpack: Not Found
    @vuepress/cli:  2.0.0-beta.35
    @vuepress/client:  2.0.0-beta.35
    @vuepress/core:  2.0.0-beta.35
    @vuepress/markdown:  2.0.0-beta.35
    @vuepress/plugin-active-header-links:  2.0.0-beta.35
    @vuepress/plugin-back-to-top:  2.0.0-beta.35
    @vuepress/plugin-container:  2.0.0-beta.35
    @vuepress/plugin-docsearch: Not Found
    @vuepress/plugin-external-link-icon:  2.0.0-beta.35
    @vuepress/plugin-git:  2.0.0-beta.35
    @vuepress/plugin-google-analytics: Not Found
    @vuepress/plugin-medium-zoom:  2.0.0-beta.35
    @vuepress/plugin-nprogress:  2.0.0-beta.35
    @vuepress/plugin-palette:  2.0.0-beta.35
    @vuepress/plugin-prismjs:  2.0.0-beta.35
    @vuepress/plugin-pwa: Not Found
    @vuepress/plugin-pwa-popup: Not Found
    @vuepress/plugin-register-components: Not Found
    @vuepress/plugin-search: Not Found
    @vuepress/plugin-shiki: Not Found
    @vuepress/plugin-theme-data:  2.0.0-beta.35
    @vuepress/plugin-toc: Not Found
    @vuepress/shared:  2.0.0-beta.35
    @vuepress/theme-default:  2.0.0-beta.35
    @vuepress/utils:  2.0.0-beta.35
    vue:  3.2.30
    vue-loader: Not Found
    vue-router:  4.0.12
    vuepress: ^2.0.0-beta.35 => 2.0.0-beta.35
    vuepress-vite:  2.0.0-beta.35
    vuepress-webpack: Not Found
```
Is this a bug or does it require any additional steps ?

**Comment by Mister-Hope** [comment, marked as answer]
naive-ui may not support ssr rendering.

There is one thing you should know: ClientOnly only render client on client side, so if the component calls client only variables in setup functions, (e.g.: `window`, `docuement`), it will still work.

But, when the component or the library itself are referencing those variables in code direct (probably context or outside component and functions), those variables will be accessed while parsing and loading that module. Also, the component is not allowed to access dom unless in `onMounted` lifecycle.

Since I did not use native ui before, you should be better sure that it's ssr friendly, otherwise, you must take actions to make a workaround.

Also make sure you are not affected by #585

**Reply by zedo75** [reply]
Thank you for your answer. My doubt is that when I have run a dev process and installed the dependency, it can work normally. However, when I restart the dev process, it reports an error. Also, is `vuepress vite` different from the `vue vite` ? I tried introduce `ElementPlus` too, but I also encountered a similar problem. [Naive UI](https://www.naiveui.com/en-US/os-theme/docs/ssr) says that its support for SSR seems feasible. So is there any possible effective solution to this?

**Reply by Mister-Hope** [reply]
Then you may be affected by #585.

**Reply by Mister-Hope** [reply]
Possiable solution, try webpack and see if it works, if so use webpack before #585 is fixed.

**Reply by Mister-Hope** [reply]
meteorlxy is busy these days, also I don't have time digging this issue, so this issue leaves for 2 month.

https://github.com/vuepress/vuepress-next/issues/585#issuecomment-1034675348

**Reply by zedo75** [reply]
Thank you very much 👍🏻 . In order to use vite happily, I'd better wait until these issues are fixed .

**Reply by Mister-Hope** [reply]
It has a workaround with #585 now, just noticing you that.
